TitaniumTask/app/src/main/java/com/sa/testtask/screens/MainActivity.java

```diff
@@ -125,35 +114,47 @@ private void update(int position) {
         Response.Image image = imageList.get(position);
         edit_label.setText(image.getName());
         Picasso.with(this)
-                .load(image.getImage())
-                .fit()
-                .into(imageView);
+            .load(image.getImage())
+            .fit()
+            .into(imageView);
     }
 
     private Subscription onSaveClick() {
         return RxView.clicks(save)
-                .map(aVoid -> edit_label.getText().toString())
-                .doOnNext(string -> imageList.get(position).setName(string))
-                .subscribe(aVoid -> {
-                    save.setVisibility(View.GONE);
-                    hideKeybord();
-                }, throwable -> Log.d(TAG, throwable.getMessage(), throwable));
+            .map(aVoid -> edit_label.getText().toString())
+            .doOnNext(string -> imageList.get(position).setName(string))
+            .subscribe(aVoid -> {
+                save.setVisibility(View.GONE);
+                hideKeybord();
+            }, throwable -> Log.d(TAG, throwable.getMessage(), throwable));
     }
 
     private Subscription onRandomClick() {
         return RxView.clicks(random)
-                .map(aVoid -> getRandomPosition())
-                .subscribe(this::update
-                        , throwable -> Log.d(TAG, throwable.getMessage(), throwable));
+            .map(aVoid -> getRandomPosition())
+            .subscribe(this::update
+                , throwable -> Log.d(TAG, throwable.getMessage(), throwable));
     }
 
     private Subscription onSelectClick() {
         return RxView.clicks(select)
-                .subscribe(aVoid -> startListActivity()
-                        , throwable -> Log.d(TAG, throwable.getMessage(), throwable));
+            .subscribe(aVoid -> startListActivity()
+                , throwable -> Log.d(TAG, throwable.getMessage(), throwable));
+    }
+
+    @Override
+    protected void onNewIntent(Intent intent) {
+        super.onNewIntent(intent);
+        //just for example
+        Log.d(TAG, "onNewIntent: got new intent");
+        position = intent.getIntExtra(IMAGE_POSITION_KEY, 0);
+        update(position);
     }
 
     private void startListActivity() {
+        //you are creating multiple instances of this activity every time user selecting image
```

Comment:
I think you are right, it really better solution. I havent thought about onActivityResult ((((